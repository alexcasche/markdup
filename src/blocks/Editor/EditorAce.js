import styled from 'styled-components'
import { tint } from 'polished'
import Ace from 'react-ace'

/*  Ace overwrites theme prop */
import appTheme from 'constants/theme'

const EditorAce = styled(Ace)`
  overflow: visible !important;
  .ace_gutter {
    background-color: transparent;
  }
  .ace_gutter-layer {
    border-right: 1px solid ${tint(0.5, appTheme.colors.grey)};
    background-color: #fafafa;
  }
  .ace_gutter-cell {
    font-size: 11px;
    font-weight: 300;
    color: #aaa;
  }
  .ace_scrollbar-v {
    display: none !important;
  }
`

export default EditorAce
