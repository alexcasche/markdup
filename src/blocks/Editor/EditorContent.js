import PreviewContent from '../Preview/PreviewContent'

const EditorContent = PreviewContent.extend`
  padding: ${props => (props.modifiers.indexOf('expanded') === -1 ? '0' : '')};
`

export default EditorContent
