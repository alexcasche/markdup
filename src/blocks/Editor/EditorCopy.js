import styled from 'styled-components'
import { tint } from 'polished'
import { applyStyleModifiers } from 'styled-components-modifiers'

const MODIFIERS = {
  visible: ({ theme }) => `
    opacity: 1;
  `,
}

const EditorCopy = styled.div`
  position: absolute;
  top: 42px;
  right: -20px;
  width: 160px;
  pointer-events: none;
  padding: 8px 16px;
  background-color: ${props => tint(0.85, props.theme.colors.black)};
  border-radius: 2px;
  color: ${props => tint(0.95, props.theme.colors.grey)};
  font-size: 14px;
  text-transform: capitalize;
  opacity: 0;
  transition: opacity 0.1s ease-in-out;
  svg {
    position: absolute;
    right: 26px;
    top: -18px;
    height: 30px;
    color: ${props => tint(0.85, props.theme.colors.black)};
  }
  ${applyStyleModifiers(MODIFIERS)};
`

export default EditorCopy
