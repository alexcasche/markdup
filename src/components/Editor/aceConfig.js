import 'brace/mode/html'
import 'brace/mode/markdown'
import 'brace/theme/github'
import 'brace/theme/chrome'
import 'brace/ext/language_tools'

const settings = {
  focus: false,
  showPrintMargin: false,
  tabSize: 2,
  useSoftTabs: true,
  vScrollBarAlwaysVisible: true,
  wrapEnabled: true,
}

const styles = {
  maxWidth: '100%',
  width: '100%',
  height: '100%',
  minHeight: '100%',
  lineHeight: '2em',
  fontFamily:
    "'Monaco', 'Menlo', 'Ubuntu Mono', 'Consolas', 'source-code-pro', monospace",
  letterSpacing: 'normal',
}

export default {
  settings,
  styles,
}
