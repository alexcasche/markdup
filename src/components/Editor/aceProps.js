const aceProps = props => {
  const {
    activeEditor,
    expanded,
    html,
    lang,
    liveComplete,
    markdown,
    splitView,
  } = props
  const mode = !splitView ? activeEditor : lang
  return {
    editorProps: { $blockScrolling: true },
    mode: mode,
    theme: mode === 'markdown' ? 'github' : 'chrome',
    value: mode === 'markdown' ? markdown : html,
    setOptions: {
      enableLiveAutocompletion: mode === 'html' ? liveComplete : false,
      fontSize: !expanded ? '13px' : '16px',
      maxLines: Infinity,
      showGutter: !expanded,
      useWorker: true,
    },
  }
}

export default aceProps
