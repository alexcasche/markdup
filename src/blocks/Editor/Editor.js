import Preview from '../Preview/Preview'
import EditorAce from './EditorAce'
import EditorContent from './EditorContent'
import EditorCopy from './EditorCopy'

const Editor = Preview.extend``

Editor.Ace = EditorAce
Editor.Content = EditorContent
Editor.Copy = EditorCopy

export default Editor
