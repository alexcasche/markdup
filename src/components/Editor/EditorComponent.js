import React, { Component } from 'react'
import { Editor } from 'blocks'
import { CopyToClipboard } from 'react-copy-to-clipboard'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
  faCaretUp,
  faCode,
  faExpandArrowsAlt,
  faEye,
  faFont,
  faPaperclip,
  faTimes,
} from '@fortawesome/free-solid-svg-icons'
import aceConfig from './aceConfig'
import aceProps from './aceProps'
import { PropTypes as t } from 'prop-types'

export default class EditorComponent extends Component {
  constructor() {
    super()
    this.state = {
      editor: undefined,
      showCopy: false,
    }
  }
  componentDidUpdate(newProps, newState) {
    const { editor, showCopy } = this.state
    if (newProps.resizing || newProps.expanded !== this.props.expanded) {
      setTimeout(() => editor.resize())
    }
    if (showCopy) {
      setTimeout(() => this.setState({ showCopy: false }), 650)
    }
  }
  copyClipboard = () => {
    this.state.editor.focus()
    this.setState({ showCopy: true })
  }
  togglePreview = () => {
    this.state.editor.focus()
    this.props.events.togglePreview()
  }
  render() {
    const {
      activeEditor,
      expanded,
      lang,
      splitView,
      events,
      actions,
    } = this.props
    const expandedModifiers = [...(expanded ? ['expanded'] : [])]
    const expandIcon = !expanded ? faExpandArrowsAlt : faTimes
    const editorIcon = activeEditor === 'markdown' ? faCode : faFont
    const nextEditor = activeEditor === 'markdown' ? 'html' : 'markdown'
    const editorTitle = !splitView ? activeEditor : lang
    const copyModifiers = [...(this.state.showCopy ? ['visible'] : [])]
    return (
      <Editor className="Editor">
        <Editor.Header modifiers={expandedModifiers}>
          <Editor.Title>{editorTitle}</Editor.Title>
          <CopyToClipboard
            text={this.props[activeEditor]}
            onCopy={this.copyClipboard}
          >
            <Editor.Button>
              <FontAwesomeIcon icon={faPaperclip} />
              <Editor.Copy modifiers={copyModifiers}>
                <FontAwesomeIcon icon={faCaretUp} />
                Copied to Clipboard!
              </Editor.Copy>
            </Editor.Button>
          </CopyToClipboard>
          {!splitView ? (
            <Editor.Button
              onClick={() =>
                actions.updateSetting({ activeEditor: nextEditor })
              }
            >
              <FontAwesomeIcon icon={editorIcon} />
            </Editor.Button>
          ) : null}
          {!expanded ? (
            <Editor.Button onClick={this.togglePreview} modifiers={['mobile']}>
              <FontAwesomeIcon icon={faEye} />
            </Editor.Button>
          ) : null}
          <Editor.Button onClick={events.toggleExpand}>
            <FontAwesomeIcon icon={expandIcon} />
          </Editor.Button>
        </Editor.Header>
        <Editor.Content modifiers={expandedModifiers}>
          <Editor.Ace
            {...aceConfig.settings}
            {...aceProps(this.props)}
            onChange={val => actions.updateWorkspace(activeEditor, val)}
            onFocus={
              splitView
                ? () => actions.updateSetting({ activeEditor: lang })
                : null
            }
            onLoad={editor => {
              editor.renderer.setPadding(10)
              this.setState({ editor: editor })
            }}
            style={aceConfig.styles}
          />
        </Editor.Content>
      </Editor>
    )
  }
}

EditorComponent.propTypes = {
  activeEditor: t.string.isRequired,
  autoSave: t.bool.isRequired,
  expanded: t.bool.isRequired,
  html: t.string.isRequired,
  lang: t.string.isRequired,
  liveComplete: t.bool.isRequired,
  markdown: t.string.isRequired,
  splitView: t.bool.isRequired,
  syncEditors: t.bool.isRequired,
  actions: t.shape({
    updateSetting: t.func.isRequired,
    updateWorkspace: t.func.isRequired,
  }).isRequired,
  events: t.shape({
    toggleExpand: t.func.isRequired,
    togglePreview: t.func.isRequired,
  }).isRequired,
}
