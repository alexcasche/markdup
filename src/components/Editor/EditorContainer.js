import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { getCurrentWorkspaceVal, getSettingVal } from 'selectors'
import { updateSetting, updateWorkspace } from 'actions'
import EditorComponent from './EditorComponent'

export const EditorContainer = props => {
  return <EditorComponent {...props} />
}

function mapStateToProps(state) {
  return {
    activeEditor: getSettingVal(state, 'activeEditor'),
    autoSave: getSettingVal(state, 'autoSave'),
    html: getCurrentWorkspaceVal(state, 'html'),
    liveComplete: getSettingVal(state, 'liveComplete'),
    markdown: getCurrentWorkspaceVal(state, 'markdown'),
    splitView: getSettingVal(state, 'splitView'),
    syncEditors: getSettingVal(state, 'syncEditors'),
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      updateSetting: bindActionCreators(updateSetting, dispatch),
      updateWorkspace: bindActionCreators(updateWorkspace, dispatch),
    },
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditorContainer)
