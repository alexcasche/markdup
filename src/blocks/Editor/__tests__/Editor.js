import Editor from '../Editor'

describe('Editor Block', () => {
  it('renders correctly', () => {
    const tree = render(<Editor />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
