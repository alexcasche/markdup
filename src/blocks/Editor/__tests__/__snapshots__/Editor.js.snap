// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Block renders correctly 1`] = `
.c0 {
  height: 100%;
  background-color: white;
  overflow: hidden;
}

<div
  className="c0"
/>
`;
